File: rng.py
"""
Shared deterministic RNG for seeded scene generation.

This is the cross-language contract behind "a seed reproduces a scene": the
web UI (web/src/lib/rng.ts) implements the exact same algorithm, so a seed
renders the identical scene in the Python GUI and in the browser. Change
nothing here without changing it there and regenerating the golden fixtures
(export_web.py --fixtures).

The generator is PCG32 (pcg32_random_r / pcg32_srandom_r from the reference
implementation, O'Neill 2014, Apache-2.0/MIT), chosen over replicating
CPython's Mersenne Twister because the whole contract then fits in ~30
published lines per language instead of chasing stdlib internals.

The derived draws are part of the contract too, and deliberately avoid
floats: integer arithmetic is bit-exact across languages with zero care
required.

    next_u32()                   raw PCG32 output
    randbelow(n)   = (next_u32() * n) >> 32        (multiply-shift; bias
                     ~n/2^32, irrelevant at the n <= 64 used here)
    randint(a, b)  = a + randbelow(b - a + 1)      (inclusive bounds)
    choice(seq)    = seq[randbelow(len(seq))]
    weighted_choice(seq, int_weights):
                     r = randbelow(sum(w)); first index whose cumulative
                     weight exceeds r

Streams here are unrelated to the stdlib `random` streams the training
scripts seed; only the seeded-scene path (test_model.make_seeded_scene and
the web export fixtures) uses SceneRandom.
"""

MASK64 = (1 << 64) - 1
MASK32 = (1 << 32) - 1

_PCG_MULT = 6364136223846793005
# Fixed stream constant (pcg32_srandom initseq). Arbitrary but part of the
# cross-language contract: the TS implementation uses the same value.
PCG_INITSEQ = 54


class SceneRandom:
    """PCG32 with the integer draw helpers scene generation needs."""

    def __init__(self, seed: int):
        # pcg32_srandom_r(initstate=seed, initseq=PCG_INITSEQ)
        self.state = 0
        self.inc = ((PCG_INITSEQ << 1) | 1) & MASK64
        self.next_u32()
        self.state = (self.state + (seed & MASK64)) & MASK64
        self.next_u32()

    def next_u32(self) -> int:
        """pcg32_random_r: 64-bit LCG step, output via xorshift + rotate."""
        old = self.state
        self.state = (old * _PCG_MULT + self.inc) & MASK64
        xorshifted = (((old >> 18) ^ old) >> 27) & MASK32
        rot = old >> 59
        return ((xorshifted >> rot) | (xorshifted << ((-rot) & 31))) & MASK32

    def randbelow(self, n: int) -> int:
        """Uniform int in [0, n) via multiply-shift."""
        return (self.next_u32() * n) >> 32

    def randint(self, a: int, b: int) -> int:
        """Uniform int in [a, b] (both bounds inclusive)."""
        return a + self.randbelow(b - a + 1)

    def choice(self, seq):
        return seq[self.randbelow(len(seq))]

    def weighted_choice(self, seq, weights):
        """Pick from seq with relative INTEGER weights (floats are banned
        from the contract)."""
        r = self.randbelow(sum(weights))
        acc = 0
        for item, w in zip(seq, weights):
            acc += w
            if r < acc:
                return item
        return seq[-1]  # unreachable when weights are positive

File: web/src/lib/rng.ts
/**
 * PCG32 + integer draw helpers: TS half of the cross-language RNG contract
 * with rng.py (see that file's docstring for the full spec). Golden
 * fixtures: web/fixtures/rng.json (tests/rng.test.ts).
 *
 * The 64-bit state lives in a BigInt. Constants are built with BigInt(...)
 * rather than the `123n` literal syntax because tsconfig targets ES2017,
 * where TypeScript rejects BigInt literals (the operators are fine).
 */

const MASK64 = BigInt("0xffffffffffffffff");
const MASK32 = BigInt("0xffffffff");
const PCG_MULT = BigInt("6364136223846793005");

const B0 = BigInt(0);
const B1 = BigInt(1);
const SHIFT_18 = BigInt(18);
const SHIFT_27 = BigInt(27);
const SHIFT_32 = BigInt(32);
const SHIFT_59 = BigInt(59);

/** Fixed stream constant (pcg32_srandom initseq); pinned by the contract. */
export const PCG_INITSEQ = 54;

export class SceneRandom {
  private state: bigint;
  private readonly inc: bigint;

  constructor(seed: number) {
    // pcg32_srandom_r(initstate=seed, initseq=PCG_INITSEQ)
    this.state = B0;
    this.inc = ((BigInt(PCG_INITSEQ) << B1) | B1) & MASK64;
    this.nextU32();
    this.state = (this.state + (BigInt(seed) & MASK64)) & MASK64;
    this.nextU32();
  }

  /** pcg32_random_r: 64-bit LCG step, output via xorshift + rotate. */
  nextU32(): number {
    const old = this.state;
    this.state = (old * PCG_MULT + this.inc) & MASK64;
    const xorshifted = Number((((old >> SHIFT_18) ^ old) >> SHIFT_27) & MASK32);
    const rot = Number(old >> SHIFT_59);
    return ((xorshifted >>> rot) | (xorshifted << ((-rot) & 31))) >>> 0;
  }

  /** Uniform int in [0, n) via multiply-shift. */
  randbelow(n: number): number {
    return Number((BigInt(this.nextU32()) * BigInt(n)) >> SHIFT_32);
  }

  /** Uniform int in [a, b], both bounds inclusive. */
  randint(a: number, b: number): number {
    return a + this.randbelow(b - a + 1);
  }

  choice<T>(seq: readonly T[]): T {
    return seq[this.randbelow(seq.length)];
  }

  /** Pick from seq with relative INTEGER weights. */
  weightedChoice<T>(seq: readonly T[], weights: readonly number[]): T {
    let total = 0;
    for (const w of weights) total += w;
    const r = this.randbelow(total);
    let acc = 0;
    for (let i = 0; i < seq.length; i++) {
      acc += weights[i];
      if (r < acc) return seq[i];
    }
    return seq[seq.length - 1]; // unreachable when weights are positive
  }
}
